File: clustering_avgcount.py
#!/usr/bin/env python3
import argparse
from pathlib import Path

import numpy as np
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score


def parse_args():
    p = argparse.ArgumentParser(description="Build LOSO cluster assignments for Model 4 (_avgcount flow).")
    p.add_argument(
        "--input",
        default="/home/runner/work/BikeCount/BikeCount/PreparedForModel_micro_nomicro.csv",
        help="Input CSV path",
    )
    p.add_argument(
        "--output",
        default="/home/runner/work/BikeCount/BikeCount/model4_outputs/clusters_avgcount.csv",
        help="Output cluster CSV path",
    )
    p.add_argument("--max-k", type=int, default=12, help="Maximum k to consider for silhouette selection")
    return p.parse_args()


MANUAL_DIRECTIONALITY_MAP = {
    "CB1101": "inbound",
    "CB1142": "inbound",
    "CB1143": "outbound",
    "CEK18": "inbound",
    "CEK049": "balanced",
    "CEE016": "balanced",
    "CB1599": "inbound",
    "CB1699": "outbound",
    "CB2105": "outbound",
    "CJM90": "inbound",
    "CEV011": "outbound",
    "CB02411": "balanced",
    "CJE181": "inbound",
    "CAT17": "outbound",
    "CLW239": "balanced",
    "COM205": "balanced",
    "CVT387": "inbound",
    "CEK31": "inbound",
}

DIR_TO_CODE = {"outbound": 0, "balanced": 1, "inbound": 2}


def choose_k_by_silhouette(x: np.ndarray, max_k: int) -> int:
    n = x.shape[0]
    if n <= 2:
        return 1
    k_upper = min(max_k, n - 1)
    if k_upper < 2:
        return 1
    best_k = 2
    best_s = -np.inf
    for k in range(2, k_upper + 1):
        km = KMeans(n_clusters=k, random_state=42, n_init=10)
        labels = km.fit_predict(x)
        if len(np.unique(labels)) < 2:
            continue
        s = silhouette_score(x, labels, metric="euclidean")
        if s > best_s:
            best_s = s
            best_k = k
    return best_k


def station_aggregate(df: pd.DataFrame, feature_cols: list[str]) -> pd.DataFrame:
    out = df.groupby("FeatureID", as_index=False)[feature_cols].mean(numeric_only=True)
    for c in feature_cols:
        out[c] = pd.to_numeric(out[c], errors="coerce")
    return out


def main():
    args = parse_args()
    in_path = Path(args.input)
    out_path = Path(args.output)
    out_path.parent.mkdir(parents=True, exist_ok=True)

    df = pd.read_csv(in_path)
    required = {
        "FeatureID",
        "Build 125m",
        "Tree Proportion",
        "Plant Proportion",
        "Grass Proportion",
    }
    missing = sorted(required - set(df.columns))
    if missing:
        raise RuntimeError(f"Missing required columns: {missing}")

    df["FeatureID"] = df["FeatureID"].astype(str)
    if "Directionality_encoded" not in df.columns:
        df["Directionality_encoded"] = 0.0
    df["Directionality_encoded"] = pd.to_numeric(df["Directionality_encoded"], errors="coerce").fillna(0.0)

    df["dir_label"] = (
        df["FeatureID"].str.upper().map(MANUAL_DIRECTIONALITY_MAP).fillna("balanced")
    )
    df["ManualFlowDirection_encoded"] = df["dir_label"].map(DIR_TO_CODE).astype(float)

    feature_cols = [
        "Build 125m",
        "Tree Proportion",
        "Plant Proportion",
        "Grass Proportion",
        "Directionality_encoded",
        "ManualFlowDirection_encoded",
    ]

    stations = sorted(df["FeatureID"].dropna().unique().tolist())
    rows = []

    for test_station in stations:
        train_df = df[df["FeatureID"] != test_station].copy()
        test_df = df[df["FeatureID"] == test_station].copy()
        if train_df.empty or test_df.empty:
            continue

        train_station = station_aggregate(train_df, feature_cols).fillna(0.0)
        x_cols = feature_cols
        x_train_raw = train_station[x_cols].astype(float).to_numpy()
        mu = np.nanmean(x_train_raw, axis=0)
        sd = np.nanstd(x_train_raw, axis=0)
        sd[~np.isfinite(sd) | (sd == 0)] = 1.0
        x_train = (np.nan_to_num(x_train_raw, nan=0.0) - mu) / sd

        k_use = choose_k_by_silhouette(x_train, max_k=max(1, int(args.max_k)))
        km = KMeans(n_clusters=max(1, k_use), random_state=42, n_init=10)
        train_labels = km.fit_predict(x_train)
        train_station["cluster_id"] = train_labels.astype(int)

        # Held-out station assignment: no held-out count leakage.
        test_static = station_aggregate(test_df, feature_cols)
        if test_static.empty:
            continue
        x_test = test_static[x_cols].astype(float).to_numpy()
        x_test = (np.nan_to_num(x_test, nan=0.0) - mu) / sd
        test_label = int(km.predict(x_test)[0])

        for _, r in train_station[["FeatureID", "cluster_id"]].iterrows():
            rows.append(
                {
                    "Fold_Test_Station": str(test_station),
                    "FeatureID": str(r["FeatureID"]),
                    "cluster_id": int(r["cluster_id"]),
                    "is_test": 0,
                }
            )
        rows.append(
            {
                "Fold_Test_Station": str(test_station),
                "FeatureID": str(test_station),
                "cluster_id": int(test_label),
                "is_test": 1,
            }
        )

    if not rows:
        raise RuntimeError("No clusters generated.")

    out = pd.DataFrame(rows)
    out.to_csv(out_path, index=False)
    print(f"Wrote clusters to {out_path}")


if __name__ == "__main__":
    main()

File: clustering_avgcount.R
#!/usr/bin/env Rscript

suppressWarnings(suppressMessages({
  library(stats)
}))

args <- commandArgs(trailingOnly = TRUE)
input_path <- if (length(args) >= 1) args[[1]] else "/home/runner/work/BikeCount/BikeCount/PreparedForModel_micro_nomicro.csv"
output_path <- if (length(args) >= 2) args[[2]] else "/home/runner/work/BikeCount/BikeCount/model4_outputs/clusters_avgcount.csv"
n_clusters <- if (length(args) >= 3) as.integer(args[[3]]) else 12L
if (is.na(n_clusters) || n_clusters < 1) n_clusters <- 12L

mean_silhouette <- function(X, labels) {
  n <- nrow(X)
  if (is.null(n) || n < 3) return(NA_real_)
  labs <- as.integer(labels)
  k <- length(unique(labs))
  if (k < 2) return(NA_real_)
  D <- as.matrix(dist(X))
  sil <- rep(NA_real_, n)
  for (i in seq_len(n)) {
    own <- labs[i]
    own_idx <- which(labs == own)
    own_idx <- own_idx[own_idx != i]
    a_i <- if (length(own_idx) == 0) 0 else mean(D[i, own_idx])
    b_i <- Inf
    for (other in setdiff(unique(labs), own)) {
      other_idx <- which(labs == other)
      if (length(other_idx) == 0) next
      b_i <- min(b_i, mean(D[i, other_idx]))
    }
    if (!is.finite(b_i)) b_i <- a_i
    denom <- max(a_i, b_i)
    sil[i] <- if (denom <= 0) 0 else (b_i - a_i) / denom
  }
  mean(sil, na.rm = TRUE)
}

choose_k_by_silhouette <- function(X, k_max) {
  n <- nrow(X)
  if (is.null(n) || n <= 2) return(1L)
  k_max <- min(as.integer(k_max), n - 1L)
  if (k_max < 2) return(1L)
  best_k <- 2L
  best_s <- -Inf
  for (k in 2:k_max) {
    set.seed(42)
    km <- kmeans(X, centers = k, iter.max = 100, nstart = 10)
    s <- mean_silhouette(X, km$cluster)
    if (!is.na(s) && s > best_s) {
      best_s <- s
      best_k <- as.integer(k)
    }
  }
  as.integer(best_k)
}

df <- read.csv(input_path, stringsAsFactors = FALSE, check.names = FALSE)

required_cols <- c(
  "FeatureID", "Count", "year_",
  "Build 125m", "Tree Proportion", "Plant Proportion", "Grass Proportion", "Directionality_encoded"
)
missing_cols <- setdiff(required_cols, names(df))
if (length(missing_cols) > 0) {
  stop(sprintf("Missing required columns: %s", paste(missing_cols, collapse = ", ")))
}

df$FeatureID <- as.character(df$FeatureID)
df$Count <- suppressWarnings(as.numeric(df$Count))
df$year_ <- suppressWarnings(as.integer(df$year_))

stations <- sort(unique(df$FeatureID))
rows_out <- list()
row_i <- 1L

for (test_station in stations) {
  train_df <- df[df$FeatureID != test_station, , drop = FALSE]
  test_df <- df[df$FeatureID == test_station, , drop = FALSE]
  if (nrow(train_df) == 0 || nrow(test_df) == 0) next

  train_feats <- aggregate(
    train_df[, c("Build 125m", "Tree Proportion", "Plant Proportion", "Grass Proportion", "Directionality_encoded")],
    by = list(FeatureID = train_df$FeatureID),
    FUN = function(x) mean(suppressWarnings(as.numeric(x)), na.rm = TRUE)
  )

  train_hist <- train_df[train_df$year_ %in% c(2022L, 2023L), c("FeatureID", "Count"), drop = FALSE]
  train_mean <- aggregate(
    train_hist$Count,
    by = list(FeatureID = train_hist$FeatureID),
    FUN = function(x) mean(suppressWarnings(as.numeric(x)), na.rm = TRUE)
  )
  names(train_mean)[2] <- "mean_count_2022_2023"

  train_station <- merge(train_feats, train_mean, by = "FeatureID", all.x = TRUE)
  train_station$mean_count_2022_2023[is.na(train_station$mean_count_2022_2023)] <- mean(train_station$mean_count_2022_2023, na.rm = TRUE)
  train_station[is.na(train_station)] <- 0

  feat_mat <- as.matrix(train_station[, c("Build 125m", "Tree Proportion", "Plant Proportion", "Grass Proportion", "Directionality_encoded", "mean_count_2022_2023"), drop = FALSE])
  feat_mat <- scale(feat_mat)
  feat_mat[is.na(feat_mat)] <- 0

  k_use <- choose_k_by_silhouette(feat_mat, n_clusters)
  if (is.na(k_use) || k_use < 1) k_use <- 1L
  set.seed(42)
  km <- kmeans(feat_mat, centers = k_use, iter.max = 100, nstart = 10)
  train_station$cluster_id <- as.integer(km$cluster) - 1L

  test_feats <- aggregate(
    test_df[, c("Build 125m", "Tree Proportion", "Plant Proportion", "Grass Proportion", "Directionality_encoded")],
    by = list(FeatureID = test_df$FeatureID),
    FUN = function(x) mean(suppressWarnings(as.numeric(x)), na.rm = TRUE)
  )
  test_hist <- test_df[test_df$year_ %in% c(2022L, 2023L), c("FeatureID", "Count"), drop = FALSE]
  if (nrow(test_hist) == 0) {
    test_mean_count <- mean(train_station$mean_count_2022_2023, na.rm = TRUE)
  } else {
    test_mean_count <- mean(suppressWarnings(as.numeric(test_hist$Count)), na.rm = TRUE)
  }
  if (is.na(test_mean_count)) test_mean_count <- mean(train_station$mean_count_2022_2023, na.rm = TRUE)
  if (nrow(test_feats) == 0) next
  test_feats$mean_count_2022_2023 <- test_mean_count
  test_feats[is.na(test_feats)] <- 0

  test_mat <- as.matrix(test_feats[, c("Build 125m", "Tree Proportion", "Plant Proportion", "Grass Proportion", "Directionality_encoded", "mean_count_2022_2023"), drop = FALSE])
  center <- attr(feat_mat, "scaled:center")
  scalev <- attr(feat_mat, "scaled:scale")
  if (is.null(center)) center <- rep(0, ncol(test_mat))
  if (is.null(scalev)) scalev <- rep(1, ncol(test_mat))
  scalev[is.na(scalev) | scalev == 0] <- 1
  test_mat <- sweep(test_mat, 2, center, "-")
  test_mat <- sweep(test_mat, 2, scalev, "/")
  test_mat[is.na(test_mat)] <- 0

  d <- as.matrix(dist(rbind(km$centers, test_mat)))
  test_cluster <- which.min(d[1:nrow(km$centers), nrow(d)]) - 1L

  train_out <- data.frame(
    Fold_Test_Station = as.character(test_station),
    FeatureID = as.character(train_station$FeatureID),
    cluster_id = as.integer(train_station$cluster_id),
    is_test = 0L,
    stringsAsFactors = FALSE
  )
  test_out <- data.frame(
    Fold_Test_Station = as.character(test_station),
    FeatureID = as.character(test_station),
    cluster_id = as.integer(test_cluster),
    is_test = 1L,
    stringsAsFactors = FALSE
  )

  rows_out[[row_i]] <- train_out
  row_i <- row_i + 1L
  rows_out[[row_i]] <- test_out
  row_i <- row_i + 1L
}

if (length(rows_out) == 0) {
  stop("No clusters generated.")
}

out <- do.call(rbind, rows_out)
dir.create(dirname(output_path), recursive = TRUE, showWarnings = FALSE)
write.csv(out, output_path, row.names = FALSE)
cat(sprintf("Wrote clusters to %s\n", output_path))
